docs/designs/CodeFlare-SDK-design-doc.md

```diff
@@ -0,0 +1,137 @@
+3. Testing: Unit tests for each API will be located within their respective submodules to ensure:
+  - Comprehensive test coverage, closely tied to the implementation, facilitating easier association between code and tests.
+  - Integration and upgrade tests that span multiple modules will be managed separately, ensuring robust validation of both individual and end-to-end functionality.
+
+The overall codebase would be structured as below:
+```
+codeflare_sdk/
+├── build/
+│   ├── requirements.txt
+│   ├── Dockerfile
+│   └── image_stream.yaml
+├── docs/
+│   ├── design_docs/
+│   ├── target_user_docs/
+├── examples/
+│   ├── guided_demos/
+│   └── additional_demos/
+├── src/
+│   └── codeflare_sdk/
+│       ├── common/
+│       |   ├── kubernetes_cluster/
+│       |   │   ├── __init__.py
+│       |   │   ├── auth.py
+│       |   │   ├── test_auth.py
+│       |   ├── kueue/
+│       |   │   ├── __init__.py
+│       |   │   ├── kueue.py
+│       |   │   └── test_kueue.py
+│       |   ├── widgets/
+│       |   │   ├── __init__.py
+│       |   │   ├── widgets.py
+│       |   │   └── test_widgets.py
+│       |   ├── utils/
+│       |   │   ├── __init__.py
+│       |   │   ├── utils.py
+│       |   │   └── test_utils.py
+│       ├── ray/
+│       │   ├── cluster/
+│       │   ├── client/
+│       │   ├── appwrapper/
+├── tests/
+│   ├── e2e/
+│   └── upgrade/
+├── poetry.lock
+└── pyproject.toml
+```
+
+### Client side best practices:
+1. The SDK should perform client-side validation of user inputs before sending them to the API server, to the extent possible. These validations should be limited to client-side checks, and the SDK is not responsible for performing validations that require server-side data. The SDK does not need to maintain the state of server-side objects. Examples of client-side validations include verifying user inputs, validating YAML schemas, and checking resource requests/limits based on available data.
+2. Any call made to the server should have a client instance passed through it. By default, the SDK should use the same client for all operations unless otherwise specified. For example, the client used to list clusters should also be used to fetch cluster details. The codebase should be designed so that these operations are either methods of the same client (e.g., a k8s client, RayJob client, TrainingClient) or the client instance is passed as an argument to the methods.
+3. Errors returned from the API server, to the extent possible, should be wrapped and handled explicitly.
+
+## Release:
+A new version of CodeFlare SDK will be released once every three weeks.
+For details on the release support matrix with other CodeFlare components, refer [here][codeflare_compatibility_matrix].
+RHOAI support matrix: CodeFlare SDK APIs fall under [Tier 2][RH_customer_API_support] support on `RHOAI` platform. This implies than an API would be support in for `n-1` major versions for a minimum of 9 months.
```

Comment:
Unless I'm missing something (which is quite possible) we don't currently have a branching strategy which ties CFSDK releases to RHOAI releases. We should confirm that CFSDK is continuously scanned for CVEs for all supported versions i.e. for 9 months after inclusion in a RHOAI release and 17 months in the case of a LTS EUS release. 
Do we have a plan in place here to be able to easily identify the packages and upgrade them in case of CVEs etc.?

Reply:
Good point, CVE fixes need to be addressed here. IIRC the downstream synk scanner and upstream Depedabots look for CVEs, and they are fixed in here. I'll check with @ChristianZaccaria on the process and come back to updating this.

Reply:
Have added a `CVE Management` section that talks about Snyk downstream and Dependabot upstream with automation around making sure that version of SDK shipped in NB is scanned throughout its lifecycle.